### PR TITLE
Kotlin serde

### PR DESCRIPTION
**Pull request by FiV0**
This PR removes the jackson deserialization and adds kotlin serde.

Todo
- [x] error handling for simple `@Serializable` annotations (revisit with some better error hierarchy) 
- [x] handling of non-ambiguous places where json-ld is an option (`systemTime` in TxKey)
- [x] make keyFn `@Serializable`
- [x] remote kotlin's `TODO()`s
- [x] snake-case vs camelCase. Where to do what.
- [x] XTDML serde

### CODE REVIEW

File: api/src/main/kotlin/xtdb/api/TransactionKey.kt [hunk 1]
```diff
@@ -1,12 +1,16 @@
...
+import kotlinx.serialization.Serializable
 import java.time.Instant
 
 private val TX_ID_KEY: Keyword = Keyword.intern("tx-id")
 private val SYSTEM_TIME_KEY: Keyword = Keyword.intern("system-time")
 
-data class TransactionKey(val txId: Long, val systemTime: Instant) : Comparable<TransactionKey>, ILookup, Seqable {
+
+@Serializable
+data class TransactionKey(val txId: Long, @Contextual val systemTime: Instant) : Comparable<TransactionKey>, ILookup, Seqable {
```
Comment (FiV0) [inline api/src/main/kotlin/xtdb/api/TransactionKey.kt]: @jarohen  I would also like to use `@Serialzable(AnySerde::class)` to encode Instant as a JsonLD type here, but kotlin doesn't like that `systemTime` can't be null.

File: api/src/main/kotlin/xtdb/api/InstantSerializer.kt
Comment (jarohen) [inline api/src/main/kotlin/xtdb/api/InstantSerializer.kt:1]: duplicate

File: api/src/main/kotlin/xtdb/api/TransactionKey.kt (hunk 1 above)
Comment (jarohen) [reply api/src/main/kotlin/xtdb/api/TransactionKey.kt]: as discussed, even in JSON-LD we can just have Instant as a string where it's unambiguous

File: api/src/main/kotlin/xtdb/api/query/Basis.kt [hunk 2]
```diff
@@ -1,16 +1,18 @@
...
-import java.util.ArrayList
 
 private val AT_TX_KEY: Keyword = Keyword.intern("at-tx")
 private val CURRENT_TIME_KEY: Keyword = Keyword.intern("current-time")
 
+@Serializable
 data class Basis(
     @JvmField val atTx: TransactionKey? = null,
-    @JvmField val currentTime: Instant? = null
+    @JvmField @Serializable(AnySerde::class) val currentTime: Instant? = null
```
Comment (jarohen) [inline api/src/main/kotlin/xtdb/api/query/Basis.kt]: as discussed, `@Contextual` here

File: api/src/main/kotlin/xtdb/api/query/Binding.kt [hunk 3]
```diff
@@ -1,7 +1,43 @@
...
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.descriptors.buildClassSerialDescriptor
+import kotlinx.serialization.encoding.Decoder
+import kotlinx.serialization.encoding.Encoder
+import kotlinx.serialization.json.*
 import xtdb.api.query.Expr.Companion.lVar
+import xtdb.api.query.Expr.Companion.param
+import xtdb.jsonIAE
+
+object BindingSerializer: KSerializer<Binding> {
```
Comment (jarohen) [inline api/src/main/kotlin/xtdb/api/query/Binding.kt]: BindingSerde

File: api/src/main/kotlin/xtdb/api/query/Expr.kt [hunk 4]
```diff
@@ -1,26 +1,147 @@
...
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.descriptors.buildClassSerialDescriptor
+import kotlinx.serialization.encoding.Decoder
+import kotlinx.serialization.encoding.Encoder
+import kotlinx.serialization.json.*
+import xtdb.ExprSerializer
+import xtdb.jsonIAE
+
+object ExprNullSerializer : KSerializer<Expr.Null> {
+    override val descriptor: SerialDescriptor = buildClassSerialDescriptor("ExprNull")
```
Comment (jarohen) [inline api/src/main/kotlin/xtdb/api/query/Expr.kt]: TODO: check what descriptors we should be returning here

File: api/src/main/kotlin/xtdb/api/query/Expr.kt [hunk 5]
```diff
@@ -1,26 +1,147 @@
...
+    @OptIn(ExperimentalSerializationApi::class)
+    override fun serialize(encoder: Encoder, value: Expr.Null) = encoder.encodeNull()
+    override fun deserialize(decoder: Decoder): Expr.Null = Expr.NULL
+}
+
+object ExprBoolSerializer: KSerializer<Expr.Bool> {
```
Comment (jarohen) [inline api/src/main/kotlin/xtdb/api/query/Expr.kt]: move within the classes themselves?

File: api/src/main/kotlin/xtdb/api/query/Query.kt [hunk 6]
```diff
@@ -1,23 +1,163 @@
...
+import kotlinx.serialization.json.*
 import xtdb.api.query.Expr.Companion.lVar
 import xtdb.api.query.Expr.Param
 import xtdb.api.query.Query.OrderDirection.ASC
 import xtdb.api.query.Query.OrderDirection.DESC
 import xtdb.api.query.Query.OrderNulls.FIRST
 import xtdb.api.query.Query.OrderNulls.LAST
+import xtdb.jsonIAE
+
+object PipelineSerializer : KSerializer<Query.Pipeline> {
```
Comment (jarohen) [inline api/src/main/kotlin/xtdb/api/query/Query.kt]: move within objs themselves?

File: api/src/main/kotlin/xtdb/api/query/QueryOptions.kt [hunk 7]
```diff
@@ -15,15 +21,16 @@ private val DEFAULT_ALL_VALID_TIME_KEY: Keyword = Keyword.intern("default-all-va
...
 
+@Serializable
 data class QueryOptions(
-    @JvmField val args: Map<String, *>? = null,
+    @JvmField val args: Map<String, @Contextual Any>? = null,
     @JvmField val basis: Basis? = null,
     @JvmField val afterTx: TransactionKey? = null,
-    @JvmField val txTimeout: Duration? = null,
-    @JvmField val defaultTz: ZoneId? = null,
+    @JvmField @Serializable(DurationSerde::class) val txTimeout: Duration? = null,
```
Comment (jarohen) [inline api/src/main/kotlin/xtdb/api/query/QueryOptions.kt]: `@Contextual`

File: api/src/main/kotlin/xtdb/api/tx/TxOps.kt [hunk 8]
```diff
@@ -46,71 +52,79 @@ sealed interface TxOp {
...
 
+@Serializable
 data class Put(
-    @JvmField val tableName: String,
-    @JvmField val doc: Map<String, *>,
-    @JvmField val validFrom: Instant? = null,
-    @JvmField val validTo: Instant? = null
+    @JvmField @SerialName("put") val tableName: String,
+    @JvmField val doc: Map<String, @Contextual Any>,
+    @JvmField @Serializable(AnySerde::class) val validFrom: Instant? = null,
```
Comment (jarohen) [inline api/src/main/kotlin/xtdb/api/tx/TxOps.kt]: `@Contextual`

File: api/src/main/kotlin/xtdb/api/tx/TxOptions.kt [hunk 9]
```diff
@@ -1,11 +1,15 @@
...
+import xtdb.InstantSerde
+import xtdb.ZoneIdSerde
 import java.time.Instant
 import java.time.ZoneId
 
+@Serializable
 data class TxOptions(
-    val systemTime: Instant? = null,
-    val defaultTz: ZoneId? = null,
+    @Serializable(InstantSerde::class) val systemTime: Instant? = null,
```
Comment (jarohen) [inline api/src/main/kotlin/xtdb/api/tx/TxOptions.kt]: `@Contextual`

**Pull request merged**